book/src/move-basics/enum-and-match.md

```diff
@@ -0,0 +1,287 @@
+
+#[test]
+fun test_segments() {
+    let _ = Segments(vector[
+        Segment::Empty,
+        Segment::String(b"hello".to_string()),
+        Segment::Special { content: b" ", encoding: 0 },
+        Segment::String(b"move".to_string()),
+        Segment::Special { content: b"21", encoding: 1 },
+    ]);
+}
+```
+
+All variants of the Segment enum share the same type – `Segment` – which allows us to create a
+homogeneous vector containing instances of different variants. This kind of flexibility is not
+achievable with structs, as each struct defines a single, fixed shape.
+
+## Pattern Matching
+
+Unlike structs, enums require special handling when it comes to accessing the inner value or
+checking the variant. We simply cannot read the inner fields of an enum using the `.` (dot) syntax,
+because we need to make sure that the value we are trying to access is the right one. For that Move
+offers _pattern matching_ syntax.
+
+> In this chapter we do not aim to cover all the features of pattern matching in Move, please, refer
+> to [Pattern Matching](/reference/pattern-matching.html) section in the Move Reference.
+
+Pattern matching allows conditioning the logic based on the _pattern_ of the value. It is performed
+using the `match` expression, followed by the matched value in parenthesis and the block of
+so-called _match arms_, defining the patten and expression to be performed if the pattern is right.
+
+Let's extend our example by adding a set of `is_variant`-like functions, so external packages can
+check the variant. Starting with `is_empty`.
+
+```move
+/// Whether this it's an `Empty` segment.
+public fun is_empty(s: &Segment): bool {
+    // Match is an expression, hence its value will be the return value of the function
+    match (s) {
+        Segment::Empty => true,
+        Segment::String(_str) => false,
+        Segment::Special { content: _, encoding: _ } => false
+    }
+}
+```
+
+The `match` keyword begins the expression, and `s` is the value being tested. Each match arm checks
+for a specific variant of the `Segment` enum. If s matches `Segment::Empty`, the function returns
+`true`; otherwise, it returns `false`.
+
+For variants with fields, we need to bind the inner structure to local variables (even if we don’t
+use them, marking unused values with `_` to compiler warnings).
+
+### Trick #1 - _any_ condition
+
+The Move compiler infers the type of the value used in a `match` expression and ensures that the
+_match arms_ are exhaustive – that is, all possible variants or values must be covered.
+
+However, in some cases—such as matching on a primitive value or a collection like a vector (we’ll
+cover that soon!) – it's not feasible to list every possible case. For these situations, match
+supports a wildcard pattern (`_`), which acts as a default arm. This arm is executed when no other
+patterns match.
+
+We can demonstrate this by simplifying our `is_empty` function and replacing the non-`Empty`
```

Comment:
Just an fyi, the reason it's recommended that we not use "we" is because it can be confusing. There was a pr recently where the approver responded to the submitter with "Can we clean up the commit history?" The submitter replied "Sure" and the PR was never updated. The approver meant can you do it, but the submitter thought it was a request for the approver's team to do it for the submitter. Anyway, maybe carry that thought forward when doing oher projects. Or not.